File: apps/web/src/api/reviews.ts
import { requestJson } from "@/api/client";

export type ReviewStatus =
  | "queued"
  | "running"
  | "completed"
  | "failed"
  | "timed_out"
  | "cancelled";

export interface ReviewState {
  review_id: string;
  document_id: string;
  review_pack_id: string;
  status: ReviewStatus;
  stage: "waiting" | "analysis" | "result_ready" | "finished";
  queued_at: string;
  started_at: string | null;
  finished_at: string | null;
  poll_after_ms: number | null;
  error: { code: string; message: string; retriable: boolean } | null;
}

export interface ReviewFinding {
  finding_id: string;
  ordinal: number;
  defect_id: string;
  severity: "critical" | "high" | "medium" | "low";
  confidence: number;
  location: {
    page: number | null;
    section_path: string[];
    block_id: string;
    table?: string | null;
    row?: number | string | null;
    column?: number | string | null;
    bbox?: Record<string, unknown> | null;
  };
  quote: string;
  problem: string;
  clarification: string;
  /**
   * Каким слоем найдено. `null` — слой неизвестен (ядро назвало проверку
   * именем, которого API не знает); интерфейс тогда ничего не пишет.
   * Сырых имён внутренних проверок здесь нет и быть не должно.
   */
  detection_layer: "rule" | "model" | "mixed" | null;
}

export interface ReviewFindings {
  review_id: string;
  items: ReviewFinding[];
  total: number;
}

/** Строка истории проверок: имя файла нужно, чтобы узнать свою проверку. */
export interface ReviewListItem {
  review_id: string;
  document_id: string;
  document_filename: string;
  status: ReviewStatus;
  queued_at: string;
  finished_at: string | null;
  findings_count: number;
}

export interface ReviewList {
  items: ReviewListItem[];
  total: number;
}

export function getReviews(signal?: AbortSignal) {
  return requestJson<ReviewList>("/api/reviews", { signal });
}

export function createReview(
  documentId: string,
  reviewPackId: string,
  idempotencyKey: string,
  signal?: AbortSignal,
) {
  return requestJson<ReviewState>("/api/reviews", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "Idempotency-Key": idempotencyKey,
    },
    body: JSON.stringify({ document_id: documentId, review_pack_id: reviewPackId }),
    signal,
  });
}

export function getReview(reviewId: string, signal?: AbortSignal) {
  return requestJson<ReviewState>(`/api/reviews/${encodeURIComponent(reviewId)}`, { signal });
}

export function getReviewFindings(reviewId: string, signal?: AbortSignal) {
  return requestJson<ReviewFindings>(
    `/api/reviews/${encodeURIComponent(reviewId)}/findings`,
    { signal },
  );
}

File: apps/api/src/docreview_api/api/schemas/reviews.py
"""Transport-only schemas for the Reviews API."""

from typing import Any, Literal

from pydantic import ConfigDict, Field

from docreview_api.api.schemas.common import ApiModel, OpaqueId, UtcDateTime

ReviewStatus = Literal["queued", "running", "completed", "failed", "timed_out", "cancelled"]
ReviewStage = Literal["waiting", "analysis", "result_ready", "finished"]


class ReviewCreateRequest(ApiModel):
    document_id: OpaqueId
    review_pack_id: OpaqueId


class ReviewPublicError(ApiModel):
    code: str
    message: str
    retriable: bool


class ReviewResponse(ApiModel):
    review_id: OpaqueId
    document_id: OpaqueId
    review_pack_id: OpaqueId
    status: ReviewStatus
    stage: ReviewStage
    queued_at: UtcDateTime
    started_at: UtcDateTime | None
    finished_at: UtcDateTime | None
    poll_after_ms: int | None = Field(default=None, ge=1000)
    error: ReviewPublicError | None = None


class FindingLocation(ApiModel):
    page: int | None
    section_path: list[str]
    block_id: str
    table: str | None = None
    row: int | str | None = None
    column: int | str | None = None
    bbox: dict[str, Any] | None = None

    model_config = ConfigDict(extra="allow")


class FindingResponse(ApiModel):
    finding_id: OpaqueId
    ordinal: int
    defect_id: str
    severity: Literal["critical", "high", "medium", "low"]
    confidence: float = Field(ge=0, le=1)
    location: FindingLocation
    quote: str
    problem: str
    clarification: str
    # Public, closed detection-layer vocabulary. Internal analyzer names from
    # `detected_by` must never cross the API boundary.
    detection_layer: Literal["rule", "model", "mixed"] | None = None


class ReviewWarning(ApiModel):
    code: str | None
    message: str


class FindingsResponse(ApiModel):
    review_id: OpaqueId
    items: list[FindingResponse]
    total: int = Field(ge=0)
    warnings: list[ReviewWarning]


class ReviewListItemResponse(ApiModel):
    """Строка истории проверок.

    Имя файла нужно, чтобы человек узнал свою проверку в списке: идентификатор
    для этого бесполезен. Диагностики ошибки здесь нет — в списке достаточно
    статуса, подробности видны на самой проверке.
    """

    review_id: OpaqueId
    document_id: OpaqueId
    document_filename: str
    status: ReviewStatus
    queued_at: UtcDateTime
    finished_at: UtcDateTime | None
    findings_count: int = Field(ge=0)


class ReviewListResponse(ApiModel):
    items: list[ReviewListItemResponse]
    total: int = Field(ge=0)
